### ISSUE TITLE
[Bug]: electron: .route() mock responses lose their status & headers when session.webRequest.onHeadersReceived is used

### ISSUE THREAD

**Issue by jtbandes** [microsoft/playwright#30495]
### Version

1.43.1

### High-level description

See minimal repro code at: https://gist.github.com/jtbandes/039be5d61760aeb0eb12389228e643d1

- This code uses `_electron.launch({...})` to run a small Electron app. It then installs a route handler using `await electronApp.firstWindow().route("https://example.com/**", ...)` which responds with `.fulfill({ json: { foo: "bar" } })`.

- The app's **renderer** process then invokes `fetch("https://example.com/foo")` and logs the response.
- When the app's **main** process uses [`session.defaultSession.webRequest.onHeadersReceived`](https://www.electronjs.org/docs/latest/api/web-request#webrequestonheadersreceivedfilter-listener) to modify response headers, **the renderer can no longer read the mocked status and headers**.

### Steps to reproduce

1. Clone `git clone https://gist.github.com/039be5d61760aeb0eb12389228e643d1.git` and enter the cloned directory
2. Run `npm install`
3. Run `node index.js`
4. Observe that the output from index.js indicates the mock response handler is called:
    ```
    $ node index.js
    [main stderr] [36923:0423/172019.226754:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
    
    fulfilling mock request: https://example.com/foo
    
    [renderer log] got response: Response 0 false {}
    [renderer log] json: {foo: bar}
    ```
6. In the electron app that opens, notice that the logged Response has `status: 0` and `ok: false`, and the headers are empty:
    <img width="718" alt="image" src="https://github.com/microsoft/playwright/assets/14237/a5295675-fb29-4f2a-99e1-b50c860d05f1">

    When this happens, the request appears in the Network tab as "Finished" but the headers are not visible:

    <img width="686" alt="image" src="https://github.com/microsoft/playwright/assets/14237/9bb2970d-4300-4b77-a87f-776b3a4ae8ee">
    <img width="637" alt="image" src="https://github.com/microsoft/playwright/assets/14237/9bdee299-8cc2-4e74-9c2b-fa4a6a4015cc">

7. Open `main.js` and remove or comment out [this section](https://gist.github.com/jtbandes/039be5d61760aeb0eb12389228e643d1#file-main-js-L13-L19) `session.defaultSession.webRequest.onHeadersReceived(...)`
8. Re-run `node index.js` and observe that `status`, `ok`, and `headers` now have the correct values:
    <img width="652" alt="image" src="https://github.com/microsoft/playwright/assets/14237/dc1f7213-4de0-45b4-9e04-f84fdf981aed">

    The request also appears with the correct status and headers in the Network tab:

    <img width="672" alt="image" src="https://github.com/microsoft/playwright/assets/14237/573d795c-42b1-4ce5-a1f9-e09a6205b3b4">

### Expected behavior

`session.webRequest.onHeadersReceived` is supposed to allow modifying response headers. It can be seen to work for real (non-mocked) requests:

<img width="764" alt="image" src="https://github.com/microsoft/playwright/assets/14237/07d192bc-34e7-43b1-ac2f-271152d9108e">

### Actual behavior

When `session.webRequest.onHeadersReceived` is combined with **mocked responses** (via `.route()`) it results in **`status: 0`, `ok: false`, and `headers: {}`**. The status and headers returned by the mock implementation are lost.

### Additional context

On rare occasions, it seems the `.route()` mock does not work at all, and the request actually goes to the network. This seems to be somehow correlated with opening the dev tools, and when it happens there is additional log output `[renderer error] Failed to load resource: net::ERR_INVALID_URL`. If you remove the `mainWindow.webContents.openDevTools()` line, it seems to stop happening. It might be interesting to investigate why this happens but it's tangential to my reported issue.

### Environment

```shell
System:
    OS: macOS 14.4.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 152.30 MB / 16.00 GB
  Binaries:
    Node: 18.18.2 - ~/.nvm/versions/node/v18.18.2/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.18.2/bin/yarn
    npm: 9.8.1 - ~/.nvm/versions/node/v18.18.2/bin/npm
    pnpm: 8.10.5 - ~/.nvm/versions/node/v18.18.2/bin/pnpm
  IDEs:
    VSCode: 1.88.1 - /usr/local/bin/code
  Languages:
    Bash: 3.2.57 - /bin/bash
  npmPackages:
    playwright: 1.43.1 => 1.43.1
```

**Comment by maximepishkov** [comment]
Same bug with webRequest.onErrorOccurred

**Comment by f-o-w-l** [comment]
Bug also occurs when using `session.defaultSession.webRequest.onBeforeSendHeaders`.

Mocked responses via `.route()` result in `status: 0, ok: false`, and `headers: {}`.

**Comment by Vere-Grey** [comment]
Same bug

**Comment by mkarp** [comment]
We're experiencing the same issue.

**Comment by veselints** [comment]
Same issue at our end. @jtbandes how did you approach the case? Do you set the response headers in some other way?

**Comment by jtbandes** [comment]
No, we just skipped the call to onHeadersReceived during the tests. 😕

**Comment by jtbandes** [comment]
@pavelfeldman any input on why this is not planned to be fixed?

**Comment by Vere-Grey** [comment]
@pavelfeldman please let us know why this was closed and what are the plans to address this big. This bug is major blocker in our automation. We are unable to mock and do proper request awaits because of it.

**Comment by jritzi** [comment]
In my case (testing a plugin for the Obsidian electron app), I wound up having to set up [@mswjs/interceptors](https://github.com/mswjs/interceptors), since this bug prevented me from using Playwright's Mock APIs.

**Comment by alexdunae** [comment]
@pavelfeldman Another vote for re-opening - this is an ongoing pain point that would be amazing to get resolved.

**Comment by f-o-w-l** [comment]
@pavelfeldman Would appreciate a justification for closing or re-opening this issue
